File: frontend/src/services/portal.ts
import axios from "axios";

import { apiClient } from "../lib/api";
import {
  clearStoredCandidateAuth,
  getStoredCandidateToken,
  storeCandidateAuth,
  storeCandidateProfile,
} from "../lib/portalAuthStorage";
import { JobOffer } from "./jobs";

const portalApiClient = axios.create({
  baseURL: import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8001",
});

portalApiClient.interceptors.request.use((config) => {
  const token = getStoredCandidateToken();
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
  }
  return config;
});

export type CandidateProfile = {
  id: string;
  first_name: string;
  last_name: string;
  email: string | null;
  phone: string | null;
  location: string | null;
  linkedin_url: string | null;
  portfolio_url: string | null;
  current_title: string | null;
  source: string;
  status: string;
  account_status: string;
  latest_cv_file_id: string | null;
  latest_cv_filename: string | null;
  latest_cv_uploaded_at: string | null;
};

export type CandidateAuthResponse = {
  access_token: string;
  token_type: "bearer";
  candidate: CandidateProfile;
};

export type CandidateRegisterPayload = {
  first_name: string;
  last_name: string;
  email: string;
  password: string;
  phone?: string;
  location?: string;
  current_title?: string;
};

export type CandidateLoginPayload = {
  email: string;
  password: string;
};

export type CandidateProfileUpdatePayload = {
  first_name?: string;
  last_name?: string;
  phone?: string;
  location?: string;
  linkedin_url?: string;
  portfolio_url?: string;
  current_title?: string;
};

export type PortalApplicationPayload = {
  first_name: string;
  last_name: string;
  email: string;
  phone?: string;
  location?: string;
  file: File;
};

export type PortalApplicationResponse = {
  candidate_id: string;
  application_id: string;
  cv_file_id: string;
  parsing_status: string;
  confidence_score: number | null;
  matching_result_ids: string[];
  message: string;
};

export type PortalApplicationStatusItem = {
  application_id: string;
  job_offer_id: string;
  job_title: string;
  company_name: string | null;
  application_status: string;
  current_stage: string | null;
  applied_at: string;
  cv_file_id: string | null;
  best_matching_score: number | null;
  recommendation: string | null;
  matching_result_ids?: string[];
};

export type PortalApplicationStatusResponse = {
  email: string;
  candidate_id: string | null;
  applications: PortalApplicationStatusItem[];
};

export async function getPublicJobs(): Promise<JobOffer[]> {
  const response = await apiClient.get<JobOffer[]>("/api/portal/jobs");
  return Array.isArray(response.data) ? response.data : [];
}

export async function getPublicJob(jobId: string): Promise<JobOffer> {
  const response = await apiClient.get<JobOffer>(`/api/portal/jobs/${jobId}`);
  return response.data;
}

export async function registerCandidate(payload: CandidateRegisterPayload): Promise<CandidateAuthResponse> {
  const response = await portalApiClient.post<CandidateAuthResponse>("/api/portal/auth/register", payload);
  storeCandidateAuth(response.data.access_token, response.data.candidate);
  return response.data;
}

export async function loginCandidate(payload: CandidateLoginPayload): Promise<CandidateAuthResponse> {
  const response = await portalApiClient.post<CandidateAuthResponse>("/api/portal/auth/login", payload);
  storeCandidateAuth(response.data.access_token, response.data.candidate);
  return response.data;
}

export async function getCandidateProfile(): Promise<CandidateProfile> {
  const response = await portalApiClient.get<CandidateProfile>("/api/portal/profile");
  storeCandidateProfile(response.data);
  return response.data;
}

export async function updateCandidateProfile(payload: CandidateProfileUpdatePayload): Promise<CandidateProfile> {
  const response = await portalApiClient.put<CandidateProfile>("/api/portal/profile", payload);
  storeCandidateProfile(response.data);
  return response.data;
}

export async function replaceCandidateCv(file: File): Promise<CandidateProfile> {
  const formData = new FormData();
  formData.append("file", file);
  const response = await portalApiClient.put<CandidateProfile>("/api/portal/profile/cv", formData, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  storeCandidateProfile(response.data);
  return response.data;
}

export async function submitAuthenticatedApplication(jobId: string): Promise<PortalApplicationResponse> {
  const response = await portalApiClient.post<PortalApplicationResponse>(`/api/portal/jobs/${jobId}/apply-auth`);
  return response.data;
}

export async function getCandidateApplications(): Promise<PortalApplicationStatusItem[]> {
  const response = await portalApiClient.get<PortalApplicationStatusItem[]>("/api/portal/applications");
  return Array.isArray(response.data) ? response.data : [];
}

export function logoutCandidate() {
  clearStoredCandidateAuth();
}

export async function submitPortalApplication(jobId: string, payload: PortalApplicationPayload): Promise<PortalApplicationResponse> {
  const formData = new FormData();
  formData.append("first_name", payload.first_name);
  formData.append("last_name", payload.last_name);
  formData.append("email", payload.email);
  if (payload.phone) {
    formData.append("phone", payload.phone);
  }
  if (payload.location) {
    formData.append("location", payload.location);
  }
  formData.append("file", payload.file);

  const response = await apiClient.post<PortalApplicationResponse>(`/api/portal/jobs/${jobId}/apply`, formData, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return response.data;
}

export async function getPortalApplicationStatus(email: string): Promise<PortalApplicationStatusResponse> {
  const response = await apiClient.get<PortalApplicationStatusResponse>("/api/portal/status", {
    params: { email },
  });
  return {
    ...response.data,
    applications: Array.isArray(response.data.applications) ? response.data.applications : [],
  };
}

File: backend/app/schemas/portal.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator


class CandidateRegister(BaseModel):
    first_name: str = Field(min_length=1, max_length=100)
    last_name: str = Field(min_length=1, max_length=100)
    email: EmailStr
    password: str = Field(min_length=8, max_length=128)
    phone: str | None = Field(default=None, max_length=50)
    location: str | None = Field(default=None, max_length=255)
    current_title: str | None = Field(default=None, max_length=150)


class CandidateLogin(BaseModel):
    email: EmailStr
    password: str = Field(min_length=1, max_length=128)


class CandidateProfileUpdate(BaseModel):
    first_name: str | None = Field(default=None, min_length=1, max_length=100)
    last_name: str | None = Field(default=None, min_length=1, max_length=100)
    phone: str | None = Field(default=None, max_length=50)
    location: str | None = Field(default=None, max_length=255)
    linkedin_url: str | None = None
    portfolio_url: str | None = None
    current_title: str | None = Field(default=None, max_length=150)


class CandidateProfileRead(BaseModel):
    id: UUID
    first_name: str
    last_name: str
    email: EmailStr | None
    phone: str | None
    location: str | None
    linkedin_url: str | None
    portfolio_url: str | None
    current_title: str | None
    source: str
    status: str
    account_status: str
    latest_cv_file_id: UUID | None = None
    latest_cv_filename: str | None = None
    latest_cv_uploaded_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class CandidateTokenResponse(BaseModel):
    access_token: str
    token_type: str = "bearer"
    candidate: CandidateProfileRead


class PortalApplicationResponse(BaseModel):
    candidate_id: UUID
    application_id: UUID
    cv_file_id: UUID
    parsing_status: str
    confidence_score: float | None
    matching_result_ids: list[UUID]
    message: str


class PortalApplicationStatusItem(BaseModel):
    application_id: UUID
    job_offer_id: UUID
    job_title: str
    company_name: str | None
    application_status: str
    current_stage: str | None
    applied_at: datetime
    cv_file_id: UUID | None
    best_matching_score: float | None
    recommendation: str | None


class PortalApplicationStatusResponse(BaseModel):
    email: EmailStr
    candidate_id: UUID | None
    applications: list[PortalApplicationStatusItem]


class CandidateApplicationRead(PortalApplicationStatusItem):
    matching_result_ids: list[UUID] = []


class PortalCandidateData(BaseModel):
    first_name: str = Field(min_length=1, max_length=100)
    last_name: str = Field(min_length=1, max_length=100)
    email: EmailStr
    phone: str | None = Field(default=None, max_length=50)
    location: str | None = Field(default=None, max_length=255)


class PublicJobRead(BaseModel):
    id: UUID
    title: str
    company_name: str | None
    location: str | None
    contract_type: str | None
    required_skills: list[str]
    preferred_skills: list[str]
    required_experience_years: int | None
    education_level: str | None
    description: str
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

    @field_validator("required_skills", "preferred_skills", mode="before")
    @classmethod
    def default_empty_skill_lists(cls, value: list[str] | None) -> list[str]:
        return value or []
